components/layout_2020/display_list/stacking_context.rs

```diff
@@ -192,25 +193,70 @@ impl DisplayList {
     }
 }
 
-pub(crate) struct StackingContextFragment {
-    scroll_node_id: ScrollTreeNodeId,
-    clip_chain_id: wr::ClipChainId,
-    section: StackingContextSection,
-    containing_block: PhysicalRect<Length>,
-    fragment: ArcRefCell<Fragment>,
+pub(crate) enum StackingContextFragment {
+    /// A normal fragment.
+    Fragment {
+        scroll_node_id: ScrollTreeNodeId,
+        clip_chain_id: wr::ClipChainId,
+        section: StackingContextSection,
+        containing_block: PhysicalRect<Length>,
+        fragment: ArcRefCell<Fragment>,
+    },
+
+    /// An index into StackingContext.inline_stacking_containers.
+    InlineStackingContainer { index: usize },
 }
 
 impl StackingContextFragment {
-    fn build_display_list(&self, builder: &mut DisplayListBuilder) {
-        builder.current_scroll_node_id = self.scroll_node_id;
-        builder.current_clip_chain_id = self.clip_chain_id;
-        self.fragment
-            .borrow()
-            .build_display_list(builder, &self.containing_block, self.section);
+    fn section(&self) -> StackingContextSection {
+        match self {
+            Self::Fragment { section, .. } => *section,
+            Self::InlineStackingContainer { .. } => StackingContextSection::Content,
+        }
+    }
+
+    fn build_display_list(
+        &self,
+        builder: &mut DisplayListBuilder,
+        inline_stacking_containers: &[StackingContext],
+    ) {
+        #[cfg(debug_assertions)]
+        let indent = "    ".repeat(builder.debug_print_indent);
+        match self {
+            Self::Fragment {
+                scroll_node_id,
+                clip_chain_id,
+                section,
+                containing_block,
+                fragment,
+            } => {
+                #[cfg(debug_assertions)]
+                {
+                    debug!("{}{:?}", indent, section);
+                }
```

Comment:
Let's remove the debug statements from this code and have a separate set of debug functions that are called when a certain log level is activated or some other mechanism. I'm not sure that `debug_assertions` is the right trigger since these aren't really assertions.

Reply:
that was the approach i used initially, with a Stacking­Context­.debug­_print method that prints the whole tree, but i decided interleaving it with the real logic would be better. having a separate method means we need to duplicate the painting order logic, so it can fall out of sync with the real logic over time and become misleading.

debug­_assertions is the same trigger that [log](https://crates.io/crates/log) uses to [compile out log levels](https://docs.rs/crate/log/0.4.20/source/src/lib.rs) in release builds, but i don’t feel too strongly about it. do you know a better trigger we can use?

Reply:
done in ff39a86828490e6f4056810850d792cc1760cd28